liquidity-amo/contracts/V3AMO.sol

```diff
@@ -0,0 +1,578 @@
+        emit UnfarmBuyBurn(
+            boostRemoved,
+            usdRemoved,
+            liquidity,
+            usdAmountIn,
+            boostAmountOut,
+            boostCollected - boostRemoved, // boostCollectedFee
+            usdCollected - usdRemoved // usdCollectedFee
+        );
+    }
+
+    ////////////////////////// PUBLIC FUNCTIONS //////////////////////////
+    function _mintSellFarm() internal override returns (uint256 liquidity, uint256 newBoostPrice) {
+        (, , , , liquidity) = _mintSellFarm(
+            uint256(type(int256).max), // boostAmount
+            1, // minBoostSpend
+            1 // minUsdSpend
+        );
+
+        newBoostPrice = boostPrice();
+    }
+
+    /**
+     * @dev _unfarmBuyBurn() is the internal function that support the the public unfarmBuyBurn() function
+     * @notice Removes liquidity, stabilizes BOOST price, and calculates the new price post-operation.
+     * @dev Fixes the issue of relying on incorrect liquidity calculations by using `quoteSwap` or Quoter Contract
+     *      to estimate the required USD amount and `_getLiquidityForUsdAmount` to determine the corresponding liquidity.
+     *      Ensures the operation only removes the necessary liquidity to achieve the target price, preventing overshooting.
+     * @return liquidity Amount of liquidity removed from the pool.
+     * @return newBoostPrice The updated BOOST price after the operation.
+     */
+    function _unfarmBuyBurn() internal override returns (uint256 liquidity, uint256 newBoostPrice) {
+        (uint256 positionLiquidity, , ) = position();
+        uint256 amountIn;
+        if (poolType == PoolType.SOLIDLY_V3) {
+            (int256 amount0, int256 amount1, , , ) = ISolidlyV3Pool(pool).quoteSwap(
+                boost > usd, // zeroForOne
+                type(int256).max,
+                targetSqrtPriceX96
+            );
+            (, int256 usdDelta) = sortAmounts(amount0, amount1);
+            amountIn = uint256(usdDelta);
+        } else if (poolType == PoolType.CL) {
+            IVeloQuoterV2.QuoteExactOutputSingleParams memory params = IVeloQuoterV2.QuoteExactOutputSingleParams({
+                tokenIn: usd,
+                tokenOut: boost,
+                amount: uint256(type(int256).max),
+                tickSpacing: IUniswapV3Pool(pool).tickSpacing(),
+                sqrtPriceLimitX96: targetSqrtPriceX96
+            });
+            (amountIn, , , ) = IVeloQuoterV2(quoter).quoteExactOutputSingle(params);
+        } else if (poolType == PoolType.ALGEBRA_V1_0 || poolType == PoolType.ALGEBRA_V1_9) {
+            (amountIn, ) = IAlgebraQuoter(quoter).quoteExactOutputSingle(
+                usd,
+                boost,
+                uint256(type(int256).max),
+                targetSqrtPriceX96
+            );
+        } else if (poolType == PoolType.ALGEBRA_INTEGRAL) {
+            IAlgebraQuoter.QuoteExactOutputSingleParams memory params = IAlgebraQuoter.QuoteExactOutputSingleParams({
```

Comment:
This will work for regular pools in AlgebraIntegral. But there are also custom pools, which would require passing the `deployer` field here (only custom pools have non-zero deployer field).

Want to double check if we are not planning on integrating with custom pools, or else we'd have to add the `deployer` field here as well.

```solidity
    struct QuoteExactInputSingleParams {
        address tokenIn;
        address tokenOut;
@>      address deployer;
        uint256 amountIn;
        uint160 limitSqrtPrice;
    }
```

https://github.com/cryptoalgebra/Algebra/blob/master/src/periphery/contracts/lens/QuoterV2.sol#L136

Reply:
Good point. We'll check it.

Reply:
@MiKoronjoo The new update here https://github.com/AXION-MONEY/liquidity-amo/blob/release/resolved_issues/contracts/V3AMO.sol#L470-L488 still passes a struct with no `deployer` field when the customDeployer doesn't exist. This call will not go through because AlgebraIntegral forces a deployer field (if it doesn't exist, it should be `address(0)`).



https://github.com/cryptoalgebra/Algebra/blob/master/src/periphery/contracts/interfaces/IQuoterV2.sol#L92-L123
```solidity

    struct QuoteExactOutputSingleParams {
        address tokenIn;
        address tokenOut;
        address deployer;
        uint256 amount;
        uint160 limitSqrtPrice;
    }

    function quoteExactOutputSingle(
        QuoteExactOutputSingleParams memory params
    )
        external
        returns (
            uint256 amountOut,
            uint256 amountIn,
            uint160 sqrtPriceX96After,
            uint32 initializedTicksCrossed,
            uint256 gasEstimate,
            uint16 fee
        );
```

You can also check they handle pool.deployer == 0 logic when computing pool address: https://github.com/cryptoalgebra/Algebra/blob/master/src/periphery/contracts/libraries/PoolAddress.sol#L41

Reply:
@pkqs90 
I checked all of the Algebral Integral DEXs that are listed in the Algebra docs:
You can see the `QuoterV2` contract code for each of them:
**Fenix**:
https://vscode.blockscan.com/81457/0x94Ca5B835186A37A99776780BF976fAB81D84ED8
**Swapsicle**:
https://mantlescan.info/address/0xBD41FEC7c74444026F8B853b4796F012750fF0Df/contract/5000/code
**Kim**:
https://modescan.io/address/0x7C5AaA464f736740156fD69171505D344855d1e5/contract/34443/code
**Blade**:
https://vscode.blockscan.com/81457/0xA405A2B8bd8FA5a111471BecC8bE56B8629Eb75A
**SilverSwap**:
https://vscode.blockscan.com/250/0x191Fd01deD8CA57C62fb19DCAbe6381653D6c60e
**Horizon**:
https://vscode.blockscan.com/59144/0xe112720bb9AC5c00B3C24CD026d74E0d69b83f06
**Glyph Exchange**:
https://scan.coredao.org/address/0x20dA24b5FaC067930Ced329A3457298172510Fe7#code
**Scribe**:
https://vscode.blockscan.com/534352/0x922c15652f3A18FBE868B38BD1E3b21E3a91c7ad

If you see, all of them, except **Scribe**, are calculating the pool address without `deployer`:
```solidity
function getPool(address tokenA, address tokenB) private view returns (IAlgebraPool) {
        return IAlgebraPool(PoolAddress.computeAddress(poolDeployer, PoolAddress.getPoolKey(tokenA, tokenB)));
    }
```

And this is for **Scribe**:
```solidity
function getPool(address deployer, address tokenA, address tokenB) private view returns (IAlgebraPool) {
        return IAlgebraPool(PoolAddress.computeAddress(poolDeployer, PoolAddress.getPoolKey(deployer, tokenA, tokenB)));
    }
```

So, we should actually support both versions. I'm not sure what the best approach is, but we'll test it and provide our solution.

Reply:
Ah, I see. Seems Algebral Integral added the `deployer` field in their v1.2, and did not exist in v1.1 or before.

- https://github.com/cryptoalgebra/Algebra/blob/v1.1-integral/src/periphery/contracts/interfaces/IQuoterV2.sol
- https://github.com/cryptoalgebra/Algebra/blob/v1.2-integral/src/periphery/contracts/interfaces/IQuoterV2.sol

Reply:
@pkqs90 
To resolve this issue, we decided to first make a low-level call to the quoter contract using the function signature for v1.2. After that, we check the call status. If it fails, we make another low-level call using the function signature for v1.1. Finally, we decode the return values to obtain the `amountIn`:
```solidity
} else if (poolType == PoolType.ALGEBRA_INTEGRAL) {
    (bool success, bytes memory data) = quoter.call(
        abi.encodeWithSignature(
            "quoteExactOutputSingle((address,address,address,uint256,uint160))",
            usd,
            boost,
            poolCustomDeployer,
            uint256(type(int256).max),
            targetSqrtPriceX96
        )
    );
    if (!success)
        (, data) = quoter.call(
            abi.encodeWithSignature(
                "quoteExactOutputSingle((address,address,uint256,uint160))",
                usd,
                boost,
                uint256(type(int256).max),
                targetSqrtPriceX96
            )
        );
    (, amountIn) = abi.decode(data, (uint256, uint256));
```